README.rst
===========
gymnasdicts
===========


.. image:: https://img.shields.io/pypi/v/gymnasdicts.svg
        :target: https://pypi.python.org/pypi/gymnasdicts

.. image:: https://img.shields.io/travis/unai/gymnasdicts.svg
        :target: https://travis-ci.com/unai/gymnasdicts

.. image:: https://readthedocs.org/projects/gymnasdicts/badge/?version=latest
        :target: https://gymnasdicts.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status


.. image:: https://pyup.io/repos/github/unai/gymnasdicts/shield.svg
     :target: https://pyup.io/repos/github/unai/gymnasdicts/
     :alt: Updates


query json dicts


* Free software: Apache Software License 2.0
* Documentation: https://gymnasdicts.readthedocs.io.


Features
--------


`gymnasdicts` is a lightweight library for querying json compatible nested-dictionaries in Python.


This package exposes one class with four functions that can be chained together so that they
follow a sql-like convention.

Query
======
`Query` is the entry point for the library, it is the data-type that
is acted on and returned by all other methods.


select
======

`select` identifies and names the nested-keys in the json-objects to be used.
The first argument is the json object to be queried.
The remaining arguments are in the keyword-arg form where:

* the values are a restricted jsonpath, where no filtering is allowed.
* the keys are user-defined variables to which the values above are assigned.

where
=====
`where` filters the results of select by value. Its arguments are lambda functions
where the argument names correspond to the variables defined in `select`.

into
====
`into` defines the shape of the output. Its only argument is lambda function
where the argument names correspond to the variables defined in `select`.

aggregate
====
`aggregate` groups a sequence of payloads by some fields and sums the rest.
The fields to be summed are defined by a single jsonpath argument.

example
=======

.. code-block:: python

    from gymnasdicts import Query

    payload = {
        "patients": [
            {"id": 1, "name": "Bob", "dob": "1982-02-05"},
            {"id": 2, "name": "Sue", "dob": "2020-02-05"},
            {"id": 3, "name": "Sam", "dob": "2020-03-06"},
            {"id": 4, "name": "Kim", "dob": "2020-03-06"},
        ],
        "prescription": [
            {"id": 1, "price": 3, "drug": "Paracetamol"},
            {"id": 2, "price": 12, "drug": "Lisinopril"},
            {"id": 3, "price": 8, "drug": "Metformin"},
        ],
        "encounters": [
            {"id": 1, "patient_id": 1, "prescription_id": 1, "doctor": "Patel"},
            {"id": 2, "patient_id": 3, "prescription_id": 2, "doctor": "Patel"},
            {"id": 3, "patient_id": 3, "prescription_id": 1, "doctor": "Francis"},
            {"id": 4, "patient_id": 4, "prescription_id": 3, "doctor": "Francis"},
        ],
    }

    q = Query(payload)
    s = q.select(
        encounter_patient_id="$.encounters[*].patient_id",
        encounter_prescription_id="$.encounters[*].prescription_id",
        doctor="$.encounters[*].doctor",
        prescription_id="$.prescription[*].id",
        patient_id="$.patients[*].id",
        patient_name="$.patients[*].name",
        dob="$.patients[*].dob",
        price="$.prescription[*].price",
    )
    w = s.where(
        lambda encounter_prescription_id, prescription_id: encounter_prescription_id
        == prescription_id,
        lambda encounter_patient_id, patient_id: encounter_patient_id == patient_id,
        lambda dob: dob > "2020-01-01",
    )
    i = w.into(
        lambda price, doctor, patient_name: {
            "doctor": doctor,
            "patients": patient_name,
            "cost": price,
        }
    )
    a = i.aggregate("$.['cost', 'patients']")
    expected = [
        {"doctor": "Patel", "patients": ["Sam"], "cost": [12]},
        {"doctor": "Francis", "patients": ["Sam", "Kim"], "cost": [3, 8]},
    ]
    assert list(a) == expected

FAQ
---

What about joins?
===================
`select` is effectively a cartesian join on all supplied jsonpaths,
i.e.

.. code-block:: python

    Query({...}).select(x="$Tbl[:].a", y="$Tbl[:].b", z="$Tbl[:].c")

is equivalent to

.. code-block:: sql

    select A.a as x, B.b as y, C.c as z from Tbl as A, Tbl as B, Tbl as C

so that `where` can be used to do the job of `on`.

This is hideous, what about memory?!
=======================================
Generators take care of this.


Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

Thanks to kclaurelie_ for useful discussion re: the relationship between select/where and keys/values.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage
.. _kclaurelie: https://github.com/kclaurelie


HISTORY.rst
=======
History
=======

0.1.0 (2020-07-28)
------------------

* First release on PyPI.

0.2.0 (2021-05-24)
------------------

* Added aggregate method.


tests/test_gymnasdicts.py
#!/usr/bin/env python

"""Tests for `gymnasdicts` package."""

import pytest


from gymnasdicts import gymnasdicts


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


gymnasdicts/tests/__init__.py
"""Unit test package for gymnasdicts."""


tests/test_gymnasdicts/test_base.py
import pytest  # type: ignore

from gymnasdicts.base import aggregate, into, select, where


@pytest.mark.parametrize(
    "payload, conditions, expected",
    [
        (
            [
                dict(x=1, y=1, z=1),
                dict(x=1, y=2, z=1),
                dict(x=2, y=1, z=1),
                dict(x=2, y=2, z=1),
            ],
            (lambda x, y: x == y,),
            [{"x": 1, "y": 1, "z": 1}, {"x": 2, "y": 2, "z": 1}],
        ),
        (
            [
                dict(x=1, y=1, z=1),
                dict(x=1, y=2, z=1),
                dict(x=2, y=1, z=1),
                dict(x=2, y=2, z=1),
            ],
            (lambda x, y: x == y, lambda x, z: x > z),
            [{"x": 2, "y": 2, "z": 1}],
        ),
    ],
)
def test_where(payload, conditions, expected):
    assert list(where(payload, *conditions)) == expected


@pytest.mark.parametrize(
    "payload, conditions, expected",
    [
        (
            [
                dict(x=1, y=1, z=1),
                dict(x=1, y=2, z=1),
                dict(x=2, y=1, z=1),
                dict(x=2, y=2, z=1),
            ],
            lambda x, y, z: x * y + z,
            [2, 3, 3, 5],
        ),
        (
            [
                dict(x=[1, 2, 3], y=1, z=1),
                dict(x=[1, 2, 3], y=2, z=2),
                dict(x=[1, 2, 3], y=1, z=3),
                dict(x=[1, 2, 3], y=2, z=-1),
            ],
            lambda x, y, z: {y: sum(xx * z for xx in x)},
            [{1: 6}, {2: 12}, {1: 18}, {2: -6}],
        ),
    ],
)
def test_into(payload, conditions, expected):
    assert list(into(payload, conditions)) == expected


@pytest.mark.parametrize("function", [where, into])
def test_where_fail(function):
    payload = [
        dict(x=1, y=1, z=1),
        dict(x=1, y=2, z=1),
        dict(x=2, y=1, z=1),
        dict(x=2, y=2, z=1),
    ]
    with pytest.raises(ValueError) as key_error:
        next(function(payload, lambda a, b: a == b))
    assert "argument-names don't match your arg-names in your payload" == str(
        key_error.value
    )


@pytest.mark.parametrize(
    "payload, conditions, expected",
    [
        (
            {
                "db": {
                    "sales": [
                        {
                            "product": "mask",
                            "days": [
                                {"number": 4, "day": "monday"},
                                {"number": 6, "day": "wednesday"},
                            ],
                        },
                        {
                            "product": "glove",
                            "days": [
                                {"number": 1, "day": "monday"},
                                {"number": 2, "day": "tuesday"},
                            ],
                        },
                    ],
                    "prices": [
                        {"product": [{"name": "mask"}, {"name": "coat"}], "cost": 55},
                        {"product": [{"name": "glove"}], "cost": 5},
                    ],
                }
            },
            {
                "product_id_1": "$.db[*].sales[*].product",
                "number_sold": "$.db[*].sales[*].days[*].number",
                "product_id_2": "$.db[*].prices[*].product[*].name",
                "product_prices": "$.db[*].prices[*].cost",
            },
            [
                {
                    "number_sold": 4,
                    "product_id_1": "mask",
                    "product_id_2": "mask",
                    "product_prices": 55,
                },
                {
                    "number_sold": 4,
                    "product_id_1": "mask",
                    "product_id_2": "coat",
                    "product_prices": 55,
                },
                {
                    "number_sold": 4,
                    "product_id_1": "mask",
                    "product_id_2": "glove",
                    "product_prices": 5,
                },
                {
                    "number_sold": 6,
                    "product_id_1": "mask",
                    "product_id_2": "mask",
                    "product_prices": 55,
                },
                {
                    "number_sold": 6,
                    "product_id_1": "mask",
                    "product_id_2": "coat",
                    "product_prices": 55,
                },
                {
                    "number_sold": 6,
                    "product_id_1": "mask",
                    "product_id_2": "glove",
                    "product_prices": 5,
                },
                {
                    "number_sold": 1,
                    "product_id_1": "glove",
                    "product_id_2": "mask",
                    "product_prices": 55,
                },
                {
                    "number_sold": 1,
                    "product_id_1": "glove",
                    "product_id_2": "coat",
                    "product_prices": 55,
                },
                {
                    "number_sold": 1,
                    "product_id_1": "glove",
                    "product_id_2": "glove",
                    "product_prices": 5,
                },
                {
                    "number_sold": 2,
                    "product_id_1": "glove",
                    "product_id_2": "mask",
                    "product_prices": 55,
                },
                {
                    "number_sold": 2,
                    "product_id_1": "glove",
                    "product_id_2": "coat",
                    "product_prices": 55,
                },
                {
                    "number_sold": 2,
                    "product_id_1": "glove",
                    "product_id_2": "glove",
                    "product_prices": 5,
                },
            ],
        ),
    ],
)
def test_select(payload, conditions, expected):
    assert list(select([payload], **conditions)) == expected


@pytest.mark.parametrize(
    "payload, pointers, message",
    [
        ([2], {}, "unexpected payload type"),
        ([{}], {"a": "$.x"}, "'x' not found in '{}'"),
    ],
)
def test_select_fail(payload, pointers, message):
    with pytest.raises(ValueError) as value_error:
        select(payload, **pointers)
    assert str(value_error.value) == message


@pytest.mark.parametrize(
    "payload, path, expected",
    [
        (
            [
                {"month": "march", "rainfall (mm)": 231},
                {"month": "march", "rainfall (mm)": 326},
                {"month": "april", "rainfall (mm)": 129},
                {"month": "april", "rainfall (mm)": 443},
                {"month": "may", "rainfall (mm)": 261},
            ],
            "$.['rainfall (mm)']",
            [
                {"month": "march", "rainfall (mm)": [231, 326]},
                {"month": "april", "rainfall (mm)": [129, 443]},
                {"month": "may", "rainfall (mm)": [261]},
            ],
        )
    ],
)
def test_aggregate(payload, path, expected):
    assert list(aggregate(payload, path)) == expected


gymnasdicts/base.py
import collections
import inspect
from itertools import product
from typing import Any, Callable, Dict, Iterator, Tuple

from gymnasdicts.utils import (
    aggregate_two_items,
    group_by,
    merge,
    parse_pointer,
    set_dict_leaf_to_list,
)

JSON = Dict[str, Any]


def select(payloads: Iterator[JSON], **pointers: str) -> Iterator[JSON]:
    """
    :example:
        >>> payload = {
        ...     "A": [
        ...         {"C": 1, "D": True, "E": "2021-01-04"},
        ...         {"C": 1, "D": False, "E": "1982-12-2"},
        ...     ],
        ...     "B": [{"F": 1, "G":.22}, {"F": 2, "G": .43}],
        ... }
        >>> list(
        ...     select(
        ...         [payload],
        ...         a="$.A[:].D",
        ...         b="$.A[:].E",
        ...         c="$.B[:].F",
        ...         d="$.B[:].G",
        ...     )
        ... )
        [{'a': True, 'b': '2021-01-04', 'c': 1, 'd': 0.22}, {'a': True, 'b': '2021-01-04', 'c': 2, 'd': 0.43}, {'a': False, 'b': '1982-12-2', 'c': 1, 'd': 0.22}, {'a': False, 'b': '1982-12-2', 'c': 2, 'd': 0.43}]
    """

    res = []

    def _select(
        _payload: JSON, _pointers: Dict[str, Tuple[str, ...]], **_values: Any
    ) -> None:
        if isinstance(_payload, dict):
            accumulated_values = dict(_values)
            pointers_for_next_level: Dict = collections.defaultdict(dict)
            for key, (head, *tail) in _pointers.items():
                if tail:
                    pointers_for_next_level[head][key] = tail
                else:
                    if head not in _payload:
                        raise ValueError(f"'{head}' not found in '{_payload}'")
                    accumulated_values[key] = _payload[head]

            for head, kw in pointers_for_next_level.items():
                _select(_payload[head], kw, **accumulated_values)

            if not pointers_for_next_level:
                res.append(accumulated_values)

        elif isinstance(_payload, (list, tuple)):
            for value in _payload:
                _select(value, _pointers, **_values)
        else:
            raise ValueError("unexpected payload type")

    parsed_pointers = {
        pointer_key: parse_pointer(pointer) for pointer_key, pointer in pointers.items()
    }
    for payload in payloads:
        _select(payload, parsed_pointers)

    return map(merge, product(*group_by(res, tuple)))


def where(payload: Iterator[JSON], *conditions: Callable) -> Iterator[JSON]:
    """
    filters an iterator of dictionaries on conditions specifies by functions
    that return a boolean and the argument names correspond to keys in the
    dictionaries.

    :param payload:
    :param conditions:
    :return:

    :example:
        >>> payload = [
        ...     dict(x=1, y=1, z=1),
        ...     dict(x=1, y=2, z=1),
        ...     dict(x=2, y=1, z=1),
        ...     dict(x=2, y=2, z=1),
        ... ]
        >>> list(where(payload, lambda x, y: x == y, lambda x, z: x > z))
        [{'x': 2, 'y': 2, 'z': 1}]
    """
    for record in payload:
        try:
            if all(
                condition(
                    *tuple(
                        record[param]
                        for param in inspect.getfullargspec(condition).args
                    )
                )
                for condition in conditions
            ):
                yield record

        except KeyError:
            raise ValueError(
                "argument-names don't match your arg-names in your payload"
            )


def into(payload: Iterator[JSON], template: Callable) -> Iterator[JSON]:
    """
    maps dictionaries to a new format via a function whoes areguments
    match the keys in the inital payload.

    :example:
        >>> payload = [
        ...     dict(x=[1, 2, 3], y=1, z=1),
        ...     dict(x=[1, 2, 3], y=2, z=2),
        ...     dict(x=[1, 2, 3], y=1, z=3),
        ...     dict(x=[1, 2, 3], y=2, z=-1),
        ... ]
        >>> list(into(payload, lambda x, y, z: {y: sum(xx * z for xx in x)}))
        [{1: 6}, {2: 12}, {1: 18}, {2: -6}]
    """
    for record in payload:
        try:
            yield template(
                *tuple(record[param] for param in inspect.getfullargspec(template).args)
            )
        except KeyError:
            raise ValueError(
                "argument-names don't match your arg-names in your payload"
            )


def aggregate(payload: Iterator[JSON], path: str) -> Iterator[JSON]:
    """
    :example:
        >>> _payload = [
        ...     {"month": "march", "rainfall (mm)": 231},
        ...     {"month": "march", "rainfall (mm)": 326},
        ...     {"month": "april", "rainfall (mm)": 129},
        ...     {"month": "april", "rainfall (mm)": 443},
        ...     {"month": "may", "rainfall (mm)": 261},
        ... ]
        >>> _path = "$.['rainfall (mm)']"
        >>> list(aggregate(_payload, _path))
        [{'month': 'march', 'rainfall (mm)': [231, 326]}, {'month': 'april', 'rainfall (mm)': [129, 443]}, {'month': 'may', 'rainfall (mm)': [261]}]

    """
    sequence = iter(payload)
    pointer = parse_pointer(path, False)

    group = next(sequence)
    set_dict_leaf_to_list(group, *pointer)

    for element in sequence:
        try:
            set_dict_leaf_to_list(element, *pointer)
            group = aggregate_two_items(group, element, *pointer)
        except AssertionError:
            yield dict(group)
            group = element
    yield group


gymnasdicts/utils.py
import collections
from itertools import groupby
from typing import Any, Callable, Dict, Iterator, List, Optional, Sequence, Tuple

from jsonpath_ng import Child, Fields, parse


def group_by(iterable: List, key: Optional[Callable] = None) -> Iterator[List]:
    """
    :example:
        >>> payload = [("b", 2, True), ("a", 1, False), ("b", 1, True)]
        >>> result = group_by(payload, lambda x: x[0])
        >>> list(result)
        [[('a', 1, False)], [('b', 2, True), ('b', 1, True)]]
    """
    for _, values in groupby(sorted(iterable, key=key), key=key):
        yield list(values)


def merge(dictionaries: Sequence[Dict]) -> Dict:
    """merge and update many dictionaries
    :example:
        >>> merge(
        ...     (
        ...         {"a": 1, "b": 2},
        ...         {"b": 3, "c": 4},
        ...         {"c": 4, "d": 5}
        ...     )
        ... )
        {'a': 1, 'b': 3, 'c': 4, 'd': 5}

    """
    return dict(collections.ChainMap(*reversed(dictionaries)))


def _pointer_to_tuple(pointer: Any) -> Tuple[Tuple[str, ...], ...]:
    """recursively flattens jsonpath-ng tree into a tuple of tuple of str
    :example:
        >>> ptr = Child(Fields("A"), Child(Fields("B"), Fields("C", "D")))
        >>> _pointer_to_tuple(ptr)
        (('A',), ('B',), ('C', 'D'))
    """
    if isinstance(pointer, Child):
        return _pointer_to_tuple(pointer.left) + _pointer_to_tuple(pointer.right)
    if isinstance(pointer, Fields):
        return (pointer.fields,)
    return tuple()


def parse_pointer(pointer_str: str, flatten=True) -> Tuple:
    """uses jsonpath-ng lib to parse various path formats
    into a standard form returning only the relevant fields
    """
    try:
        pointer = parse(pointer_str)
    except Exception as exception:  # not my fault,
        # it really is implemented as a raw Exception in this package
        raise ValueError(f"{exception}")

    nested_tuples = _pointer_to_tuple(pointer)
    if not flatten:
        return nested_tuples
    return sum(nested_tuples, tuple())


def aggregate_two_items(left: Any, right: Any, path: Dict) -> Any:
    """
    :example:
        >>> a = {"a": {"x": [3], "y": 5, "z": 9}}
        >>> b = {"a": {"x": [4], "y": 5, "z": 10}}
        >>> aggregate_two_items(a, b, {"a": ["x", "z"]})
        {'a': {'x': [3, 4], 'y': 5, 'z': 19}}

    """

    assert type(left) == type(right)

    if isinstance(left, list):  # apply to each item
        assert len(left) == len(right)
        return [
            aggregate_two_items(left_item, right_item, path)
            for left_item, right_item in zip(left, right)
        ]

    if isinstance(left, dict):
        assert left.keys() == right.keys()
        assert set(path).issubset(left), "path contains fields not in payload!"

        ret = {}
        for key in left:
            if key not in path:  # check equal and move on
                ret[key] = aggregate_two_items(left[key], right[key], {})
            elif isinstance(path, (list, tuple)):  # end of the line, sum these objects
                ret[key] = left[key] + right[key]
            else:  # ..continue
                ret[key] = aggregate_two_items(left[key], right[key], path[key])

        return ret

    assert not path, "cant use jsonpath on a primitive!"
    assert left == right  # check objects are the same
    return left  # return one of them


def set_dict_leaf_to_list(dictionary, *keys: Tuple[str, ...]) -> None:
    """
    :example:
        >>> d = {"a": {"b": {"c": None}}}
        >>> set_dict_leaf_to_list(d, ["a"], ["b"], ["c"])
        >>> d
        {'a': {'b': {'c': [None]}}}
    """
    head, *tail = keys

    if tail:
        for item in head:
            set_dict_leaf_to_list(dictionary[item], *tail)
    else:
        for item in head:
            dictionary[item] = [dictionary[item]]


gymnasdicts/__init__.py
from __future__ import annotations

"""Top-level package for gymnasdicts."""

__author__ = """George Burton"""
__email__ = "george.burton@unai.com"
__version__ = "0.2.2"


from typing import Callable, Iterator, Union

from gymnasdicts import base


class Query:
    """

    :example:
    >>> payload = {
    ...     "patients": [
    ...         {"id": 1, "name": "Bob", "dob": "1982-02-05"},
    ...         {"id": 2, "name": "Sue", "dob": "2020-02-05"},
    ...         {"id": 3, "name": "Sam", "dob": "2020-03-06"},
    ...         {"id": 4, "name": "Kim", "dob": "2020-03-06"},
    ...     ],
    ...     "prescription": [
    ...         {"id": 1, "price": 3, "drug": "Paracetamol"},
    ...         {"id": 2, "price": 12, "drug": "Lisinopril"},
    ...         {"id": 3, "price": 8, "drug": "Metformin"},
    ...     ],
    ...     "encounters": [
    ...         {"id": 1, "patient_id": 1, "prescription_id": 1, "doctor": "Patel"},
    ...         {"id": 2, "patient_id": 3, "prescription_id": 2, "doctor": "Patel"},
    ...         {"id": 3, "patient_id": 3, "prescription_id": 1, "doctor": "Francis"},
    ...         {"id": 4, "patient_id": 4, "prescription_id": 3, "doctor": "Francis"},
    ...     ],
    ... }

    >>> q = Query(payload)
    >>> s = q.select(
    ...     encounter_patient_id="$.encounters[*].patient_id",
    ...     encounter_prescription_id="$.encounters[*].prescription_id",
    ...     doctor="$.encounters[*].doctor",
    ...     prescription_id="$.prescription[*].id",
    ...     patient_id="$.patients[*].id",
    ...     patient_name="$.patients[*].name",
    ...     dob="$.patients[*].dob",
    ...     price="$.prescription[*].price",
    ... )
    >>> w = s.where(
    ...     lambda encounter_prescription_id, prescription_id: encounter_prescription_id
    ...     == prescription_id,
    ...     lambda encounter_patient_id, patient_id: encounter_patient_id == patient_id,
    ...     lambda dob: dob > "2020-01-01",
    ... )
    >>> i = w.into(
    ...     lambda price, doctor, patient_name: {
    ...         "doctor": doctor,
    ...         "patients": patient_name,
    ...         "cost": price,
    ...     }
    ... )
    >>> a = i.aggregate("$.['cost', 'patients']")
    >>> expected = [
    ...     {"doctor": "Patel", "patients": ["Sam"], "cost": [12]},
    ...     {"doctor": "Francis", "patients": ["Sam", "Kim"], "cost": [3, 8]}
    ... ]
    >>> list(a) == expected
    True
    """

    def __init__(self, json_data: Union[base.JSON, Iterator[base.JSON]]) -> None:
        if isinstance(json_data, Iterator):
            self.json_data = json_data
        else:
            self.json_data = iter([json_data])

    def select(self, **pointers: str) -> Query:
        return Query(base.select(self.json_data, **pointers))

    def where(self, *conditions: Callable) -> Query:
        return Query(base.where(self.json_data, *conditions))

    def into(self, template: Callable) -> Query:
        return Query(base.into(self.json_data, template))

    def aggregate(self, pointer: str) -> Query:
        return Query(base.aggregate(self.json_data, pointer))

    def __iter__(self) -> Iterator[base.JSON]:
        for item in self.json_data:
            yield item
